addons/hr_expense/migrations/10.0.2.0/post-migration.py

```diff
@@ -0,0 +1,114 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Le Filament (<https://le-filament.com>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from openupgradelib import openupgrade
+from odoo import api, fields, models, _
+
+@openupgrade.migrate()
+def migrate(env, version):
+    cr = env.cr
+    # Create new expense sheet for each expense registered
+    cr.execute(
+        '''INSERT INTO hr_expense_sheet
+        (account_move_id, bank_journal_id, company_id, create_date, 
+	create_uid, currency_id, department_id, employee_id, journal_id, 
+	message_last_post, name, state, total_amount, write_date, write_uid)
+        SELECT
+         account_move_id, bank_journal_id, company_id, create_date, 
+	 create_uid, currency_id, department_id, employee_id, journal_id, 
+	 message_last_post, name, state, total_amount, write_date, write_uid
+	FROM hr_expense
+	''')
+
+    # Set sheet_id for each expense_id
+    cr.execute(
+	'''UPDATE hr_expense 
+	SET sheet_id = subquery.sheet_id
+	FROM (
+            SELECT 
+		hes.id AS sheet_id, he.id AS expense_id
+	    FROM hr_expense he, hr_expense_sheet hes
+	    WHERE he.name=hes.name and he.create_date=hes.create_date
+	) AS subquery
+	WHERE id=subquery.expense_id
+	''')
+
+    # Set address for each employee
+    cr.execute(
+	'''UPDATE hr_expense_sheet
+	SET address_id = subquery.address_home_id
+	FROM (
+	    SELECT
+		address_home_id, id
+	    FROM hr_employee
+	) AS subquery
+	WHERE employee_id = subquery.id
+	''')
+
+    # Set responsible for each expense by setting to department manager when department is filled in
+    cr.execute(
+	'''UPDATE hr_expense_sheet
+	SET responsible_id = subquery.manager_id
+	FROM (
+	    SELECT
+		manager_id, id
+	    FROM hr_department
+	) AS subquery
+	WHERE department_id = subquery.id
+	''')
+
+    # Set correct reported state for hr_expense
+    cr.execute(
+	'''UPDATE hr_expense
+	SET state='reported'
+	WHERE state='submit' or state='approved'
+	''')
+
+    # Set correct done state for hr_expense
+    cr.execute(
+	'''UPDATE hr_expense
+	SET state='done'
+	WHERE state='post' or state='done'
+	''')
+
+    # Set correct refused state for hr_expense
+    cr.execute(
+	'''UPDATE hr_expense
+	SET state='refused'
+	WHERE state='cancel'
+	''')
+  
+    # Set correct submit state for hr_expense_sheet since draft does not exist anymore for sheet
+    cr.execute(
+	'''UPDATE hr_expense_sheet
+	SET state='submit'
+	WHERE state='draft'
+	''')
+
+    # Set correct account corresponding to product subject to this expense 
+    cr.execute("SELECT distinct he.product_id, pp.product_tmpl_id from hr_expense he, product_product pp where he.product_id=pp.id")
+    product_ids = cr.fetchall()
+    for product in product_ids:
```

Comment:
Initially I tried to call a function to get the account associated to each product from the product addon (which retrieves it from ir_property table) but did not manage to figure out how to do that...
That would have been much cleaner I agree.

Regarding the specifics of your comment, I am not sure about what you propose since the account is not linked to the company, there are so many accounts per company. However the account is linked to the product template and therefore should be retrieved from there in my view...